### PR TITLE
Fixes restore of a shrunken index when initial recovery node is gone

### LINKED ISSUE #24257: Snapshot and restore and shrink do not play well together
The simplest way to reproduce locally is with `gradle run` and then:

```
POST /test/doc
{ "test": "test" }


PUT /test/_settings
{
  "settings": {
    "index.routing.allocation.require._name": "shrink_node_name", 
    "index.blocks.write": true 
  }
}

POST /test/_shrink/shrunk

PUT /_snapshot/test_repo
{
  "type": "fs",
  "settings": {
    "compress": true,
    "location": "/Users/manybubbles/Workspaces/Elasticsearch/master/elasticsearch/distribution/build/cluster/shared/repo/test_repo"
  }
}

POST /_snapshot/test_repo/test_snapshot?wait_for_completion=true

POST /_snapshot/test_repo/test_snapshot/_restore?wait_for_completion=true
{
  "indices": ["shrunk"]
}
```

Kibana will give you "socket hang up" error because Elasticsearch has tripped an assertion and killed itself:
```
[elasticsearch] java.lang.AssertionError: all settings must have been upgraded before
[elasticsearch]         at org.elasticsearch.cluster.metadata.MetaDataIndexUpgradeService.upgradeIndexMetaData(MetaDataIndexUpgradeService.java:77) ~[elasticsearch-6.0.0-alpha1-SNAPSHOT.jar:6.0.0-alpha1-SNAPSHOT]
```

If you run without `-ea` then the shrunken index won't allocation properly because of the `index.routing.allocation.initial_recovery` setting. That setting cannot be cleared.

### PR DESCRIPTION
**Pull request by abeyad**
When an index is shrunk using the shrink APIs, the shrink operation adds
some internal index settings to the shrink index, for example
`index.shrink.source.name|uuid` to denote the source index, as well as
`index.routing.allocation.initial_recovery._id` to denote the node on
which all shards for the source index resided when the shrunken index
was created.  However, this presents a problem when taking a snapshot of
the shrunken index and restoring it to a cluster where the initial
recovery node is not present, or restoring to the same cluster where the
initial recovery node is offline or decommissioned.  The restore
operation fails to allocate the shard in the shrunken index to a node
when the initial recovery node is not present, and a restore type of
recovery will *not* go through the PrimaryShardAllocator, meaning that
it will not have the chance to force allocate the primary to a node in
the cluster.  Rather, restore initiated shard allocation goes through
the BalancedShardAllocator which does not attempt to force allocate a
primary.

This commit fixes the aforementioned problem by not requiring allocation
to occur on the initial recovery node when the recovery type is a
restore of a snapshot.  This commit also ensures that the internal
shrink index settings are recognized and not archived (which can trip an
assertion in the restore scenario).

Closes #24257

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/snapshots/DedicatedClusterSnapshotRestoreIT.java [hunk 1]
```diff
@@ -801,33 +798,71 @@ public void run() {
...
+     * Tests that a shrunken index (created via the shrink APIs) and subsequently snapshotted
+     * can be restored when the node the shrunken index was created on is no longer part of
+     * the cluster.
+     */
+    public void testRestoreShrinkIndex() throws Exception {
+        logger.info("-->  starting a master node and a data node");
+        internalCluster().startMasterOnlyNode();
+        internalCluster().startDataOnlyNode();
 
-    private boolean snapshotIsDone(String repository, String snapshot) {
```
Comment (abeyad) [inline core/src/test/java/org/elasticsearch/snapshots/DedicatedClusterSnapshotRestoreIT.java]: removing dead code here

File: core/src/main/java/org/elasticsearch/cluster/routing/RecoverySource.java [hunk 2]
```diff
@@ -250,7 +250,7 @@ public String toString() {
         }
     }
 
-    private static EnumSet<RecoverySource.Type> INITIAL_RECOVERY_TYPES = EnumSet.of(Type.EMPTY_STORE, Type.LOCAL_SHARDS, Type.SNAPSHOT);
+    private static EnumSet<RecoverySource.Type> INITIAL_RECOVERY_TYPES = EnumSet.of(Type.EMPTY_STORE, Type.LOCAL_SHARDS);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/RecoverySource.java]: @ywelsch can you look at this one please

File: core/src/main/java/org/elasticsearch/common/settings/IndexScopedSettings.java [hunk 3]
```diff
@@ -191,9 +191,15 @@ protected boolean isPrivateSetting(String key) {
             case IndexMetaData.SETTING_VERSION_UPGRADED:
             case IndexMetaData.SETTING_INDEX_PROVIDED_NAME:
             case MergePolicyConfig.INDEX_MERGE_ENABLED:
+            case IndexMetaData.INDEX_SHRINK_SOURCE_UUID_KEY:
+            case IndexMetaData.INDEX_SHRINK_SOURCE_NAME_KEY:
                 return true;
             default:
-                return false;
+                if (key.startsWith(IndexMetaData.INDEX_ROUTING_INITIAL_RECOVERY_GROUP_SETTING.getKey())) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/settings/IndexScopedSettings.java]: maybe use `return IndexMetaData.INDEX_ROUTING_INITIAL_RECOVERY_GROUP_SETTING.getRawKey().match(key)`

File: core/src/main/java/org/elasticsearch/cluster/routing/RecoverySource.java (hunk 2 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/cluster/routing/RecoverySource.java]: @abeyad leave a comment why we dont' have snapshot in here?

**Review by s1monw — APPROVED** [review]
I left some comments but LGTM I want @ywelsch  to look at it as well

**Review by ywelsch — CHANGES_REQUESTED** [review]
I've left a comment about the change of INITIAL_RECOVERY_TYPES. I am also debating whether a better solution would be to keep the definition as is but to automatically remove the `initial_recoveries` settings during restore (see `ignoreSettings` in `RestoreService.updateIndexSettings`).

**Comment by abeyad** [comment]
@s1monw @ywelsch I pushed a6d83302d71e9e769ad4830725902b1f84f3f85b to address the feedback.

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/FilterAllocationDecider.java [hunk 4]
```diff
@@ -75,6 +77,17 @@
...
 
+    /**
+     * The set of {@link RecoverySource.Type} values for which the
+     * {@link IndexMetaData#INDEX_ROUTING_INITIAL_RECOVERY_GROUP_SETTING} should apply.
+     * Note that we do not include the {@link RecoverySource.Type#SNAPSHOT} type here
+     * because if the snapshot is restored to a different cluster that does not contain
+     * the initial recovery node id, or to the same cluster where the initial recovery node
+     * id has been decommissioned, then the primary shards will never be allocated.
+     */
+    public static EnumSet<RecoverySource.Type> INITIAL_RECOVERY_TYPES =
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/FilterAllocationDecider.java]: no need to make this public; package-visible is good enough.

**Review by ywelsch — APPROVED** [review]
LGTM

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/FilterAllocationDecider.java (hunk 4 above)
Comment (abeyad) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/FilterAllocationDecider.java]: the problem is that `FilterAllocationDeciderTests` is not part of the same package, and moving it to the same package as `FilterAllocationDecider` breaks package visibility on the `AllocationService#reroute` protected method that the test depends on.

Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/FilterAllocationDecider.java]: just use the public one with one less parameter...

**Comment by abeyad** [comment]
@ywelsch I pushed 6ad6dea0039b2a91b78b6df23407042ad8acb04c

**Comment by abeyad** [comment]
5.x commit: 30ed3baf58936e1de8db7fda7568cd7b3f645db0
5.4 commit: 4d7aaf4179251b20df502096516d626b7add39cf

**Comment by abeyad** [comment]
thanks for the review @ywelsch @s1monw

**Pull request merged**